Implement js.lib.WeakRef

dimensionscape opened the pull request:
WeakRef was introduced in ES2021 and is now supported on every major browser. I think it's time for Haxe to include it as well.

dimensionscape committed:
Implement js.lib.WeakRef

WeakRef was introduced in ES6 and is now supported on every major browser. I think it's time for Haxe to include it as well.

dimensionscape committed:
Use Type parameter for WeakRef

haxiomic commented:
Minor tweak, maybe it's worth adding the constraint `{}`: `class WeakRef<T: {}>` as you can only have weak refs to objects and it's a runtime error to ref primitives like Int or String

WeakRef was introduced in ES2021 so it's quite new but still widely supported (~87% https://caniuse.com/?search=weakref) so I agree it's fair for it to be in js.lib

dimensionscape commented:
> Minor tweak, maybe it's worth adding the constraint `{}`: `class WeakRef<T: {}>` as you can only have weak refs to objects and it's a runtime error to ref primitives like Int or String

Absolutely correct. I actually didn't notice that I forgot to add the constraint when applying the Type param changes to my original commit. Thanks!

> WeakRef was introduced in ES2021 so it's quite new but still widely supported (~87% https://caniuse.com/?search=weakref) so I agree it's fair for it to be in js.lib

Correct again. My apologies - I hadn't even had my morning coffee yet at the time of writing. Perhaps I was conflating with WeakMaps, but I'm not sure why I said ES6.

dimensionscape committed:
Add Type constraint to WeakRef

As Haxiomic pointed out, I forgot the Type constraint required to prevent primitives from being used which would result in a runtime error.

haxiomic commented:
All good :], thank you for the contribution <3
